Userland/Libraries/LibAudio/PlaybackStreamSerenity.cpp

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (c) 2023, Gregory Bertilson <zaggy1024@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "PlaybackStreamSerenity.h"
+
+#include <LibCore/ThreadedPromise.h>
+
+namespace Audio {
+
+ErrorOr<NonnullRefPtr<PlaybackStream>> PlaybackStreamSerenity::create(OutputState initial_state, u32 sample_rate, u8 channels, u32 target_latency_ms, AudioDataRequestCallback&& data_request_callback)
+{
+    VERIFY(data_request_callback);
+    auto connection = TRY(ConnectionToServer::try_create());
+    if (auto result = connection->try_set_self_sample_rate(sample_rate); result.is_error())
+        return Error::from_string_literal("Failed to set sample rate");
+    (void)channels, (void)target_latency_ms;
+
+    auto polling_timer = TRY(Core::Timer::try_create());
+    auto implementation = TRY(adopt_nonnull_ref_or_enomem(new (nothrow) PlaybackStreamSerenity(connection, move(polling_timer), move(data_request_callback))));
+    if (initial_state == OutputState::Playing)
+        connection->async_start_playback();
+    return implementation;
+}
+
+PlaybackStreamSerenity::PlaybackStreamSerenity(NonnullRefPtr<ConnectionToServer> stream, NonnullRefPtr<Core::Timer> polling_timer, AudioDataRequestCallback&& data_request_callback)
+    : m_connection(move(stream))
+    , m_polling_timer(move(polling_timer))
+    , m_data_request_callback(move(data_request_callback))
+{
+    // Ensure that our audio buffers are filled when they are more than 3/4 empty.
+    // FIXME: Add an event to ConnectionToServer track the sample rate and update this interval.
+    //        Or use a shared condition variable to wake an enqueuing thread to provide data to the audio server.
+    m_polling_timer->set_interval(static_cast<int>((AUDIO_BUFFERS_COUNT * 3 / 4) * AUDIO_BUFFER_SIZE * 1000 / m_connection->get_self_sample_rate()));
+    m_polling_timer->on_timeout = [this]() {
+        fill_buffers();
+    };
+    m_polling_timer->start();
+}
+
+void PlaybackStreamSerenity::fill_buffers()
+{
+    // The number below is picked arbitrarily.
+    // FIXME: This shouldn't need to poll. Ideally, we should be able to recieve a signal when the server
+    //        consumes a buffer and enqueue some data automatically.
```

Comment:
Same story here. The whole point is that there are no callbacks about consumed buffers, since those provide questionable benefit over just running a check every so often, given that the buffer "consumption rate" is highly predictable.